architecture/2024-10-deploymenttemplate-controller.md

```diff
@@ -0,0 +1,421 @@
+type DeploymentTemplatePhrase string
+
+const (
+  // DeploymentTemplatePhraseUpdating indicates that the Deployment Template is being updated.
+  DeploymentTemplatePhraseUpdating DeploymentTemplatePhrase = "Updating"
+
+  // DeploymentTemplatePhraseReady indicates that the Deployment Template is ready.
+  DeploymentTemplatePhraseReady DeploymentTemplatePhrase = "Ready"
+
+  // DeploymentTemplatePhraseFailed indicates that the Deployment Template has failed.
+  DeploymentTemplatePhraseFailed DeploymentTemplatePhrase = "Failed"
+
+  // DeploymentTemplatePhraseDeleting indicates that the Deployment Template is being deleted.
+  DeploymentTemplatePhraseDeleting DeploymentTemplatePhrase = "Deleting"
+
+  // DeploymentTemplatePhraseDeleted indicates that the Deployment Template has been deleted.
+  DeploymentTemplatePhraseDeleted DeploymentTemplatePhrase = "Deleted"
+)
+```
+
+#### DeploymentResource Custom Resource Definition
+
+The `DeploymentResource` CRD is another CRD that will be responsible for tracking the state of individual resources. The CRD will have the following fields:
+
+```go
+// DeploymentResourceSpec defines the desired state of a Deployment Resource.
+type DeploymentResourceSpec struct {
+  // Scope is the resource ID of the scope.
+	Scope string `json:"scope,omitempty"`
+
+  // ResourceId is the Radius resource Id.
+  ResourceId string `json:"resourceId"`
+}
+
+type DeploymentTemplateResourceStatus struct {
+  // Scope is the resource ID of the scope.
+	Scope string `json:"scope,omitempty"`
+
+  // ObservedGeneration is the most recent generation observed for this DeploymentTemplateResource.
+  ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+  // Operation tracks the status of an in-progress provisioning operation.
+  Operation *ResourceOperation `json:"operation,omitempty"`
+
+  // Phrase indicates the current status of the Deployment Template Resource.
+  Phrase DeploymentTemplateResourcePhrase `json:"phrase,omitempty"`
+
+  // Message is a human-readable description of the status of the Deployment Template Resource.
+  Message string `json:"description,omitempty"`
+}
+
+// DeploymentResourcePhrase is a string representation of the current status of a Deployment Resource.
+type DeploymentResourcePhrase string
+
+const (
+	// DeploymentResourcePhraseReady indicates that the Deployment Resource is ready.
+	DeploymentResourcePhraseReady DeploymentResourcePhrase = "Ready"
+
+	// DeploymentResourcePhraseFailed indicates that the Deployment Resource has failed to delete.
+	BicepPhraseFailed DeploymentResourcePhrase = "Failed"
+
+	// DeploymentResourcePhraseDeleting indicates that the Deployment Resource is being deleted.
+	BicepPhraseDeleting DeploymentResourcePhrase = "Deleting"
+
+	// DeploymentResourcePhraseDeleted indicates that the Deployment Resource has been deleted.
+	BicepPhraseDeleted DeploymentResourcePhrase = "Deleted"
+)
+```
```

Comment:
Why no updating for this resource?  what is the state when it's being created/deployed but not yet 'ready'?

Reply:
DeploymentResource is only responsible for its own deletions, not create/update